meta-lmp-base/recipes-support/lmp-el2go-auto-register/lmp-el2go-auto-register/lmp-el2go-auto-register

```diff
@@ -76,20 +83,59 @@ class Pkcs11:
                         pass  # already removed
         return len(missing) == 0
 
+    @classmethod
+    def _init_slot_dictionary(cls, pkcs11_output: str) -> dict:
+        slot_dict = {}
+        slot_id = None
+        r = re.compile("\((?P<slot_id>[0-9x]{3})\)")
+
+        for line in pkcs11_output.splitlines():
+            if line.startswith("Slot"):
+                # decode slot ID
+                # assume output in format:
+                # Slot 0 (0x0): 710be08c-e795-597d-bf30-2a98bab41050
+                g = r.search(line)
+                slot_id = g.group("slot_id")
+                slot_dict.update({slot_id: {}})
+            if line.startswith(" "):
+                # decode slot dictionary
+                # assume output in format:
+                #   token label        : testtoken
+                parts = line.split(":", 1)
+                slot_dict[slot_id].update({parts[0].strip(): parts[1].strip()})
+        return slot_dict
+
     @classmethod
     def _is_initialized(cls) -> bool:
-        args = ["pkcs11-tool", f"--module={cls.LIB}", f"--pin={PIN}", "--list-objects"]
+        args = ["pkcs11-tool", f"--module={cls.LIB}", "--list-slots"]
         try:
-            subprocess.check_output(args, stderr=subprocess.STDOUT)
+            out = subprocess.check_output(args, stderr=subprocess.STDOUT)
+            slot_dict = cls._init_slot_dictionary(out.decode())
+            for slot_id, slot in slot_dict.items():
+                token_label = slot.get("token label")
+                if token_label == f"{TOKEN_LABEL}":
+                    return True
         except subprocess.CalledProcessError:
             return False
-        return True
+        return False
 
     @classmethod
     def _initialize(cls):
+        args = ["pkcs11-tool", f"--module={cls.LIB}", "--list-slots"]
+        out = subprocess.check_output(args, stderr=subprocess.STDOUT)
+        slot_dict = cls._init_slot_dictionary(out.decode())
+        empty_slot = None
+        for slot_id, slot in slot_dict.items():
+            token_state = slot.get("token state")
+            if token_state == "uninitialized":
+                empty_slot = slot_id
+                break
+        if empty_slot is None:
+            # raise exception as there are no more available slots
+            raise Pkcs11Error("No free PKCS11 slots")
```

Comment:
I don't think you are gaining anything from a custom error class. Seems like you can just throw a RuntimeError .

Other than that - I think this is beyond my understanding of PKCS and you should have the platform team review

Reply:
+1 for a generic error here.

Reply:
done